extensions/control-plane/transfer/transfer-pull-http-dynamic-receiver/src/main/java/org/eclipse/edc/connector/receiver/http/dynamic/HttpDynamicEndpointDataReferenceReceiver.java

```diff
@@ -0,0 +1,163 @@
+import org.eclipse.edc.spi.types.domain.edr.EndpointDataReference;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import static dev.failsafe.Failsafe.with;
+import static java.lang.String.format;
+import static org.eclipse.edc.util.string.StringUtils.isNullOrBlank;
+
+/**
+ * Implementation of a {@link EndpointDataReferenceReceiver} that posts
+ * the {@link EndpointDataReference} to an existing http endpoint sent by clients when initiating the data transfer.
+ */
+public class HttpDynamicEndpointDataReferenceReceiver implements EndpointDataReferenceReceiver {
+
+
+    public static final String HTTP_RECEIVER_ENDPOINT = "receiver.http.endpoint";
+
+    private static final MediaType JSON = MediaType.get("application/json");
+
+    private Monitor monitor;
+    private OkHttpClient httpClient;
+    private TypeManager typeManager;
+    private RetryPolicy<Object> retryPolicy;
+
+    private TransferProcessStore transferProcessStore;
+
+    private String authKey;
+    private String authToken;
+
+    private HttpDynamicEndpointDataReferenceReceiver() {
+    }
+
+    @Override
+    public CompletableFuture<Result<Void>> send(@NotNull EndpointDataReference edr) {
+
+        var processId = transferProcessStore.processIdForDataRequestId(edr.getId());
+
+        if (processId == null) {
+            throw new EdcException(format("Failed to found processId for DataRequestId %s", edr.getId()));
+        }
+        var transferProcess = transferProcessStore.find(processId);
+
+        if (transferProcess == null) {
+            throw new EdcException(format("Failed to found transfer process for id %s", processId));
+        }
+
+        var endpoint = transferProcess.getProperties().get(HTTP_RECEIVER_ENDPOINT);
+
+        if (endpoint != null) {
+            return sendEdr(edr, endpoint);
+        } else {
+            monitor.debug(format("Missing %s property in the transfer process properties", HTTP_RECEIVER_ENDPOINT));
+            return CompletableFuture.completedFuture(Result.success());
+        }
+
+
+    }
+
+    @NotNull
+    private CompletableFuture<Result<Void>> sendEdr(@NotNull EndpointDataReference edr, String endpoint) {
+
+
+        var requestBody = RequestBody.create(typeManager.writeValueAsString(edr), JSON);
+        var requestBuilder = new Request.Builder().url(endpoint).post(requestBody);
+        if (!isNullOrBlank(authKey) && !isNullOrBlank(authToken)) {
+            requestBuilder.header(authKey, authToken);
+        }
+        try (var response = with(retryPolicy).get(() -> httpClient.newCall(requestBuilder.build()).execute())) {
+            if (response.isSuccessful()) {
+                var body = response.body();
+                if (body == null) {
+                    throw new EdcException(format("Received empty response body when receiving endpoint data reference at uri: %s", endpoint));
+                }
```

Comment:
Why is this a problem?

Reply:
It has the same behavior if the fixed http receiver

https://github.com/eclipse-edc/Connector/blob/a2995869307c2f1a1609017209be71e3688e71c7/extensions/control-plane/transfer/transfer-pull-http-receiver/src/main/java/org/eclipse/edc/connector/receiver/http/HttpEndpointDataReferenceReceiver.java#L66-L68

Reply:
For the response i agree it did make sense to fail in this case.

we should change also `HttpEndpointDataReferenceReceiver`

Reply:
I think this is because an our misuse of the okhttp client. As the `Response.body` documentation reports:
```
Returns a non-null value if this response was passed to Callback.onResponse or returned from Call.execute. Response bodies must be closed and may be consumed only once.
This always returns null on responses returned from cacheResponse, networkResponse, and priorResponse.
```
In our case that object will always be non-null, so, ok to check it just to avoid the warning, but if we don't have to fetch the body I would remove that control completely.